mono/sgen/sgen-gc.c

```diff
@@ -1819,7 +1819,12 @@ major_copy_or_mark_from_roots (SgenGrayQueue *gc_thread_gray_queue, size_t *old_
 			 * We force the finish of the worker with the new object ops context
 			 * which can also do copying. We need to have finished pinning.
 			 */
-			sgen_workers_start_all_workers (object_ops, NULL);
+			/* FIXME Implement parallel copying and get rid of this ineffective hack */
+			if (major_collector.is_parallel)
+				sgen_workers_start_all_workers (&major_collector.major_ops_conc_par_start, NULL);
+			else
+				sgen_workers_start_all_workers (object_ops, NULL);
```

Comment:
At this point in the implementation, only the parallel concurrent context is properly implemented, since we don't have yet parallel copying (that is done later in the PR). Here we are doing a forced finish having as a context (normal object_ops) the non-parallel concurrent_finish context. The forced finish means that first each worker clears it's leftovers from the private gray queues, which can only be done with a context that supports parallelism.